docs/proposals/20210323-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,556 @@
+
+##### Contract between etcd provider and controlplane provider
+- Controlplane providers should check for the presence of the paused annotation (`cluster.x-k8s.io/paused`), and not continue provisioning if it is set. The KubeadmControlPlane controller does that.
+- Controlplane providers should for presence of `Cluster.Spec.ManagedExternalEtcdRef` field. This check should happen after the controlplane is unpaused and before it starts provisioning.
+- The controlplane provider should "Get" the CR referred by the cluster.spec.ManagedExternalEtcdRef and check its status.endpoint field. It will parse the endpoints from this field and use these endpoints where applicable. For instance, the KCP controller will read the status.endpoint field and parse it into a string slice, and then use it as the external etcd endpoints.
+- The controlplane provider will read the etcd CA cert, and the client cert-key pair from the two Kubernetes Secrets named `{cluster.Name}-apiserver-etcd-client` and `{cluster.Name}-etcd`
+
+##### Contract between etcd provider and infrastructure provider
+- Each etcd member requires ports 2379 and 2380 for client and peer communication respectively. 
+- In case of the infrastructure providers that use security groups, either the infrastrucutre provider will add a new security group specifically for etcd members, or document that user needs to create a security group allowing these two ports and use that for the etcd machines.
+
+![etcd-controller-contract](images/managed-etcd/etcd-controller-contract.png)
+---
+
+
+**Etcdadm based etcd provider**
+
+This etcd provider will have two main components:
+
+##### Etcdadm bootstrap provider
+
+- The Etcdadm bootstrap provider will convert a Machine to an etcd member by generating cloud-init scripts that will run the required etcdadm commands. It will do so through the EtcdadmConfig resource.
+- Etcdadm bootstrap provider controller will follow the same flow as that of the kubeadm bootstrap provider
+    - This controller will also use the [InitLock](https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L62) way of determining which member runs the init command in case multiple EtcdadmConfig resources are created at the same time. Along with this, the controller will also check the ManagedEtcdInitialized condition on the ClusterStatus. It will be set by the machine controller after the first node is provisioned. If it is not set, the Machine corresponding to the EtcdadmConfig resource being processed will run `etcdadm init` command. If the ManagedEtcdInitialized condition is true, then the Machine will run `etcdadm join` command. 
+     - For the init node, the controller will lookup existing CA cert-key pair, or generate a CA cert key and save them in a Secret in the management cluster. For subsequent etcd members, the controller will lookup the CA cert key pair.
```

Comment:
What is controller here? Is that etcd-manager built on top of CAPI? And how does lookup work? Via CAPI or some other external component?